src/meds_etl/flat.py

```diff
@@ -477,6 +587,201 @@ def convert_flat_to_meds(
+    conn.sql(f"SET temp_directory = '{temp_dir}';")
+    # See https://duckdb.org/docs/configuration/pragmas.html#memory-limit
+    conn.sql(f"SET memory_limit = '{memory_limit_GB}GB';")
+
+    if os.path.exists(os.path.join(source_flat_path, "metadata.json")):
+        shutil.copyfile(
+            os.path.join(source_flat_path, "metadata.json"), os.path.join(target_meds_path, "metadata.json")
+        )
+
+    csv_tasks = []
+    parquet_tasks = []
+
+    source_flat_data_path = os.path.join(source_flat_path, "flat_data")
+
+    for flat_file in os.listdir(source_flat_data_path):
+        full_flat_file = os.path.join(source_flat_data_path, flat_file)
+        if full_flat_file.endswith(".csv") or full_flat_file.endswith(".csv.gz"):
+            csv_tasks.append(full_flat_file)
+        else:
+            parquet_tasks.append(full_flat_file)
+
+    random.seed(3422342)
+    random.shuffle(csv_tasks)
+    random.shuffle(parquet_tasks)
+
+    all_views = []
+
+    metadata_columns_set = set()
+
+    tasks = []
+    all_columns = []
+    for task in csv_tasks:
+        all_columns.append(get_csv_columns(task))
+        tasks.append(task)
+
+    for task in parquet_tasks:
+        all_columns.append(get_parquet_columns(task))
+        tasks.append(task)
+
+    metadata_columns_set = {b for a in all_columns for b in a}
+
+    metadata_columns_set -= KNOWN_COLUMNS
+    metadata_columns = sorted(list(metadata_columns_set))
+
+    for i, (task, columns) in enumerate(zip(tasks, all_columns)):
+        select_parts = []
+
+        select_parts.append("cast(time as timestamp) as time")
+        select_parts.append("cast(patient_id as int64) as patient_id")
+        select_parts.append("cast(code as string) as code")
+
+        if "value" not in columns:
+            select_parts.append("cast(datetime_value as timestamp) as datetime_value")
+            select_parts.append("cast(numeric_value as float4) as numeric_value")
+            select_parts.append("cast(text_value as string) as text_value")
+        else:
+            select_parts.append("try_cast(value as timestamp) as datetime_value")
+
+            select_parts.append(
+                """
+                try_cast(
+                    (case when (try_cast(value as timestamp) is null) then value else null end)
+                    as float4
+                ) as numeric_value
+            """
+            )
+
+            select_parts.append(
+                """
+                cast(
+                    (
+                        case when (
+                            (try_cast(value as timestamp) is null ) AND (try_cast(value as float4) is null)
+                        ) then value else null end
+                    )
+                as string) as text_value
+            """
+            )
+
+        for column in metadata_columns:
+            if column in columns:
+                select_parts.append(f'cast("{column}" as string) as "{column}"')
```

Comment:
I've done some surgery to `get_parquet_columns` so that it returns both the column name and the column data type. I've also added an assertion to verify that each metadata column is of type `string` or `large_string`